### PR TITLE
[GEOS-9474] Json-ld backward mapping will not work when pointing to c…

### PR DESCRIPTION
**Pull request by taba90**
…ql function comprising xpath

<Include a few sentences describing the overall goals for this Pull Request>

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

For all pull requests:

- [X] Confirm you have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/master/CONTRIBUTING.md) 
- [X] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in documentation)
- [X] Make sure the first PR targets the master branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.

The following are required only for core and extension modules (they are welcomed, but not required, for community modules):
- [X] There is a ticket in Jira describing the issue/improvement/feature (a notable exemptions is, changes not visible to end users)
- [X] PR for bug fixes and small new features are presented as a single commit
- [X] Commit message must be in the form "[GEOS-XYZW] Title of the Jira ticket" (export to XML in Jira generates the message in this exact form)
- [X] New unit tests have been added covering the changes
- [X] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [X] This PR passes the [QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html) (QA checks results will be reported by travis-ci after opening this PR)
- [ ] Commits changing the UI, existing user workflows, or adding new functionality, need to include documentation updates (screenshots, text)
- [ ] Committs changing the REST API, or any configuration object, should check it the REST API docs (Swagger YAML files and classic documentation) need to be updated.

Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or inapplicable.

### CODE REVIEW

**Comment by taba90** [comment]
Related JIRA ticket: https://osgeo-org.atlassian.net/browse/GEOS-9474

**Comment by taba90** [comment]
depends on https://github.com/geotools/geotools/pull/2774

File: src/community/json-ld/src/main/java/org/geoserver/jsonld/builders/impl/DynamicValueBuilder.java [hunk 1]
```diff
@@ -96,57 +112,48 @@ private Object evaluateExpressions(JsonBuilderContext context) {
...
     }
 
-    private void prepareXpathFilter(Expression expr) {
-        List<Expression> params = ((Function) expr).getParameters();
-        if (params.size() > 0) {
-            for (Expression e : params) {
-                if (e instanceof XPathFunction) ((XPathFunction) e).setNamespaces(namespaces);
-                else if (e instanceof Function) {
-                    prepareXpathFilter(e);
+    private void literalXpathToPropertyName(Expression expr, String literalXpath) {
```
Comment (nmco) [inline src/community/json-ld/src/main/java/org/geoserver/jsonld/builders/impl/DynamicValueBuilder.java]: This function would deserve some comments explaining its behavior, and ideally should be break into small chunks. No need to change, just sharing my thoughts.

File: src/community/json-ld/src/main/java/org/geoserver/jsonld/builders/impl/DynamicValueBuilder.java [hunk 2]
```diff
@@ -96,57 +112,48 @@ private Object evaluateExpressions(JsonBuilderContext context) {
...
-        // extract xpath from cql expression if present
-        int xpathI = expression.indexOf("xpath(");
-        if (xpathI != -1) {
-            int xpathI2 = expression.indexOf(")", xpathI);
-            String xpath = expression.substring(xpathI, xpathI2 + 1);
-            determineContextPos(xpath);
-            expression = expression.replaceAll("\\.\\./", "");
+        if (expr instanceof FunctionExpression) {
+            ((FunctionExpression) expr).setParameters(newParams);
+        } else if (expr instanceof MathExpressionImpl) {
```
Comment (nmco) [inline src/community/json-ld/src/main/java/org/geoserver/jsonld/builders/impl/DynamicValueBuilder.java]: Here you probably want to check if its a binary expression and setting it accordingly no?

File: src/community/json-ld/src/main/java/org/geoserver/jsonld/builders/impl/DynamicValueBuilder.java [hunk 3]
```diff
@@ -38,10 +41,23 @@ public DynamicValueBuilder(String key, String expression, NamespaceSupport names
         super(key);
         this.namespaces = namespaces;
         if (expression.startsWith("$${")) {
-            this.cql = ExpressionsUtils.extractCqlExpressions(workXpathFunction(expression));
+            // takes xpath fun from cql
```
Comment (nmco) [inline src/community/json-ld/src/main/java/org/geoserver/jsonld/builders/impl/DynamicValueBuilder.java]: This should be isolated in a more meaningful private method and some comments explaining waht's going on added.

File: src/community/json-ld/src/main/java/org/geoserver/jsonld/builders/impl/DynamicValueBuilder.java [hunk 4]
```diff
@@ -38,10 +41,23 @@ public DynamicValueBuilder(String key, String expression, NamespaceSupport names
...
+            // clean the function to obtain a cql expression without xpath() syntax
+            this.cql =
+                    extractCqlExpressions(
+                            cleanCQLExpression(expression, strXpathFun, literalXpath));
+            // replace the xpath literal inside the expression with a PropertyName
+            literalXpathToPropertyName(this.cql, removeBackDots(literalXpath));
         } else if (expression.startsWith("${")) {
-            String strXpath = ExpressionsUtils.extractXpath(expression);
-            strXpath = determineContextPos(strXpath);
+            String strXpath = extractXpath(expression);
```
Comment (nmco) [inline src/community/json-ld/src/main/java/org/geoserver/jsonld/builders/impl/DynamicValueBuilder.java]: Same as above here.

File: src/community/json-ld/src/main/java/org/geoserver/jsonld/expressions/ExpressionsUtils.java [hunk 5]
```diff
@@ -220,6 +222,12 @@ public static String quoteXpathAttribute(String xpath) {
         return xpath;
     }
 
+    /**
+     * Extract Namespaces from given FeatureType
+     *
+     * @param type
+     * @return
+     */
```
Comment (nmco) [inline src/community/json-ld/src/main/java/org/geoserver/jsonld/expressions/ExpressionsUtils.java]: This java doc looks bad.

File: src/community/json-ld/src/main/java/org/geoserver/jsonld/expressions/ExpressionsUtils.java [hunk 6]
```diff
@@ -251,4 +259,70 @@ public static String removeQuotes(String cqlFilter) {
...
+     * encoded as a PropertyName
+     *
+     * @param expression
+     * @param toReplace
+     * @param replacement
+     * @return
+     */
+    public static String cleanCQLExpression(
+            String expression, String toReplace, String replacement) {
+        if (expression.indexOf(XPATH_FUN_START) != -1)
```
Comment (nmco) [inline src/community/json-ld/src/main/java/org/geoserver/jsonld/expressions/ExpressionsUtils.java]: @taba90 I have a doubt here, if someone writes ``xpath  (`` (notice the extra spaces) in a CQL expression will this be valid?

File: src/community/json-ld/src/main/java/org/geoserver/jsonld/builders/impl/DynamicValueBuilder.java (hunk 1 above)
Comment (taba90) [reply src/community/json-ld/src/main/java/org/geoserver/jsonld/builders/impl/DynamicValueBuilder.java]: Thanks for the review @nmco this feedback has been applied

File: src/community/json-ld/src/main/java/org/geoserver/jsonld/builders/impl/DynamicValueBuilder.java (hunk 3 above)
Comment (taba90) [reply src/community/json-ld/src/main/java/org/geoserver/jsonld/builders/impl/DynamicValueBuilder.java]: applied too

File: src/community/json-ld/src/main/java/org/geoserver/jsonld/builders/impl/DynamicValueBuilder.java (hunk 2 above)
Comment (taba90) [reply src/community/json-ld/src/main/java/org/geoserver/jsonld/builders/impl/DynamicValueBuilder.java]: Checking for MathExpressionImpl cause setParameters is available only casting to it. Anyway right now MathExpressionImpl is extended by all the implementation of BinaryExpression

File: src/community/json-ld/src/main/java/org/geoserver/jsonld/expressions/ExpressionsUtils.java (hunk 6 above)
Comment (taba90) [reply src/community/json-ld/src/main/java/org/geoserver/jsonld/expressions/ExpressionsUtils.java]: checked: an "xpath (" won't be considered valid

File: src/community/json-ld/src/main/java/org/geoserver/jsonld/expressions/ExpressionsUtils.java (hunk 5 above)
Comment (taba90) [reply src/community/json-ld/src/main/java/org/geoserver/jsonld/expressions/ExpressionsUtils.java]: fixed

File: src/community/json-ld/src/main/java/org/geoserver/jsonld/builders/impl/DynamicValueBuilder.java (hunk 2 above)
Comment (nmco) [reply src/community/json-ld/src/main/java/org/geoserver/jsonld/builders/impl/DynamicValueBuilder.java]: Interesting I didn't know that, thx @taba90!

File: src/community/json-ld/src/main/java/org/geoserver/jsonld/expressions/ExpressionsUtils.java (hunk 6 above)
Comment (nmco) [reply src/community/json-ld/src/main/java/org/geoserver/jsonld/expressions/ExpressionsUtils.java]: ok, thx

**Comment by aaime** [comment]
@nmco good to merge?

**Comment by nmco** [comment]
Looks good to me, this needs to be backported to 2.16.x right @taba90 ?

**Pull request merged**